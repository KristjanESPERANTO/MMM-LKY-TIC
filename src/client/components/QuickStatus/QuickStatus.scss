@import "../../styles/common";

/** @define quickstatus */

.mmm-lky-tic-wrapper {
  .quick-status {
    display: flex;
    flex-direction: column;
    padding-bottom: 0.5em;
    margin-bottom: 0.5em;
    border-bottom: 1px solid var(--color-text-dimmed);

    &__items {
      display: flex;
      flex-flow: row wrap;
    }

    &__link-item,
    &__supply-item,
    &__fare-period-item {
      margin: 0 0.5em;
    }

    &__link-picto.not-connected,
    &__supply-picto.supply-critical {
      animation: blinker 1s linear infinite;
    }

    &__fare-period-item {
      font-size: 0.8em;
    }
  }
}
