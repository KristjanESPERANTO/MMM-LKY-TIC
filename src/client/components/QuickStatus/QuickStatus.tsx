import { faCircleQuestion } from '@fortawesome/free-regular-svg-icons';
import { faBoltLightning, faLeaf, faPlugCircleBolt, faPlugCircleExclamation } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import classNames from 'classnames';
import timeDifferenceInSeconds from 'date-fns/differenceInSeconds';
import { FunctionComponent } from 'react';
import { TeleInfo } from '../../../shared/domain/teleinfo';

import './QuickStatus.scss';

export interface QuickStatusProps {
  data?: TeleInfo;
  hearbeatTs?: number;
}

/**
 * Teleinfo quick status bar
 */
const QuickStatus: FunctionComponent<QuickStatusProps> = (props: QuickStatusProps) => {
  function renderItems() {
    return (
      <>
        <li key="qs-link" className="quick-status__link-item">
          {renderLinkItem()}
        </li>
        <li key="qs-supply" className="quick-status__supply-item">
          {renderSupplyItem()}
        </li>
      </>
    );
  }

  function renderLinkItem() {
    const options = {
      color: 'red',
      icon: faPlugCircleExclamation,
      isCritical: false,
    };

    if(props.data) {
      const lastReceivedTime = props.data.meta?.lastUpdateTimestamp || 0;
      const heartbeatTime = props.hearbeatTs || 0;
      const idleDuration = timeDifferenceInSeconds(heartbeatTime, lastReceivedTime);
      
      if (idleDuration > 30) {
        options.color = 'red';
        options.icon = faPlugCircleExclamation;
        options.isCritical = true;
      } else if (idleDuration > 10) {
        options.color = 'orange';
        options.icon = faPlugCircleExclamation;
      } else if (idleDuration > 5) {
        options.color = 'yellow';
        options.icon = faPlugCircleExclamation;
      } else {
        options.color = 'green';
        options.icon = faPlugCircleBolt;  
      }
   }

    const itemClassName = classNames('quick-status__link-picto', {
      'not-connected': options.isCritical,
    });
    return (
      <>
        <FontAwesomeIcon className={itemClassName} color={options.color} icon={options.icon} />
      </>
    );
  }

  function renderSupplyItem() {
    const options = {
      color: 'white',
      icon: faCircleQuestion,
      isCritical: false,
    };

    const instantIntensity = props.data?.instantIntensity || 0;
    const maxIntensity = props.data?.subscribedIntensity || Number.POSITIVE_INFINITY;

    if (props.data?.subscribedPowerOverflowWarning) {
      options.color = 'red';
      options.icon = faBoltLightning;
      options.isCritical = true;
    } else if (instantIntensity > maxIntensity * 5 / 6) {
      options.color = 'red';
      options.icon = faBoltLightning;
    } else if (instantIntensity > maxIntensity * 2 / 3) {
      options.color = 'orange';
      options.icon = faBoltLightning;
    } else if (instantIntensity > maxIntensity / 3) {
      options.color = 'yellow';
      options.icon = faBoltLightning;
    } else if (props.data) {
      options.color = 'green';
      options.icon = faLeaf;
    }

    const itemClassName = classNames('quick-status__supply-picto', { 'supply-critical': options.isCritical });
    return (
      <>
        <FontAwesomeIcon className={itemClassName} color={options.color} icon={options.icon} />
      </>
    );
  }

  return (
    <div className="quick-status">
      <ul className="quick-status__items">
        {renderItems()}
      </ul>
    </div>
  )
};

export default QuickStatus;
