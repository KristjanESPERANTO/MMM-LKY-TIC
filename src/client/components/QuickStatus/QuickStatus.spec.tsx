import renderer from 'react-test-renderer';
import QuickStatus, { QuickStatusProps } from '../QuickStatus/QuickStatus';

describe('QuickStatus component', () => {
  const defaultProps: QuickStatusProps = {};
  const propsWithData: QuickStatusProps = {
    ...defaultProps,
    data: {
      ...defaultProps.data,
      meta: {
        firstDataTimestamp: 1674565891995,
        lastUpdateTimestamp: 1674818126026,
        unresolvedGroups: {},
      },
      apparentPower: 250,
      chosenFareOption: 'BASE',
      estimatedPower: 225,
      estimatedPrices:{
        currentDay: 5.5,
        currentMonth: 55,
        total: 500
      },
      instantIntensity: 1,
      suppliedPower: {
        currentDay: [1, 2],
        currentMonth: [10, 20],
        total: [100, 200],
      },
    },
  };

  it('should render correctly without data: no teleinfo data after more than 30 seconds', () => {
    // given-when
    const tree = renderer.create(<QuickStatus {...defaultProps} />).toJSON();

    // then
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with data', () => {
    // given-when
    const tree = renderer
      .create(<QuickStatus {...propsWithData} />)
      .toJSON();

    // then
    expect(tree).toMatchSnapshot();
  });
});
