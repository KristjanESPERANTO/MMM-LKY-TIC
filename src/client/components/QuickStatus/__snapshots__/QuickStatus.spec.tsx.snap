// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuickStatus component should render correctly with data 1`] = `
<div
  className="quick-status"
>
  <ul
    className="quick-status__items"
  >
    <li
      className="quick-status__link-item"
    >
      <div>
        FontAwesomeIcon component 
        {"className":"quick-status__link-picto not-connected","color":"red","icon":"faPlugCircleExclamation"}
      </div>
    </li>
    <li
      className="quick-status__supply-item"
    >
      <div>
        FontAwesomeIcon component 
        {"className":"quick-status__supply-picto","color":"green","icon":"faLeaf"}
      </div>
    </li>
  </ul>
</div>
`;

exports[`QuickStatus component should render correctly without data: no teleinfo data after more than 30 seconds 1`] = `
<div
  className="quick-status"
>
  <ul
    className="quick-status__items"
  >
    <li
      className="quick-status__link-item"
    >
      <div>
        FontAwesomeIcon component 
        {"className":"quick-status__link-picto","color":"red","icon":"faPlugCircleExclamation"}
      </div>
    </li>
    <li
      className="quick-status__supply-item"
    >
      <div>
        FontAwesomeIcon component 
        {"className":"quick-status__supply-picto","color":"white","icon":"faCircleQuestion"}
      </div>
    </li>
  </ul>
</div>
`;
